lib/router/src/lib.rs

```diff
@@ -208,43 +246,86 @@ where
     fn not_recognized() -> Self {
         Self::error(error::NotRecognized.into())
     }
-
-    fn no_capacity(capacity: usize) -> Self {
-        Self::error(error::NoCapacity(capacity).into())
-    }
 }
 
-impl<Req, Svc> Future for ResponseFuture<Req, Svc>
+impl<Req, Rec, Mk> Future for ResponseFuture<Req, Rec, Mk>
 where
-    Svc: svc::Service<Req>,
-    Svc::Error: Into<error::Error>,
+    Rec: Recognize<Req>,
+    Mk: Make<Rec::Target>,
+    Mk::Value: svc::Service<Req> + Clone,
+    <Mk::Value as svc::Service<Req>>::Error: Into<error::Error>,
 {
-    type Item = <LoadShed<Svc> as svc::Service<Req>>::Response;
+    type Item = <LoadShed<Mk::Value> as svc::Service<Req>>::Response;
     type Error = error::Error;
 
     fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
         use svc::Service;
 
         loop {
             self.state = match self.state {
-                State::Init(ref mut req, ref mut svc) => {
-                    match svc.poll_ready().map_err(error::Error::from)? {
+                State::Inserting(ref mut rr) => {
+                    // Aquire the lock for the router cache
+                    let mut cache = match rr.cache.poll_lock() {
+                        Async::Ready(aquired) => aquired,
+                        Async::NotReady => return Ok(Async::NotReady),
+                    };
+
+                    let request = rr.request.take().expect("polled after ready");
+                    let target = rr.target.take().expect("polled after ready");
+
+                    // If the target is already cached, route the request to
+                    // the service; otherwise, try to insert it
+                    if let Some(service) = cache.access(&target) {
+                        State::Calling(Some(request), Some(service))
+                    } else {
+                        // Ensure that there is capacity for a new slot
+                        cache.at_capacity()?;
```

Comment:
I would expect something called at_capacity/is_at_capacity to return only a boolean indicating whether there is capacity in the cache; and I still think it would be clearer if the error logic was handled here rather than in the cache.

What do you think about changing the cache to export `pub fn can_insert(&self) -> bool`; and then having this code just be `if !cache.can_insert() { return Err(NoCapacity) }` -- I don't think there's any value to returning `self.capacity`...

Reply:
My argument was that having this method return `Result` made the callsite simpler, since it wasn't responsible for constructing the error, which feels like the cache's responsibility to me. In that case, though, I wouldn't call it `at_capacity` --- `try_reserve` or something would make sense for a method returning `Result`.

Regardless, I'm fine with @olix0r's suggestion instead; this isn't a big deal to me.

Reply:
This would be fine if called `try_reserve`, imo, as well... I could go either way on this.